### ISSUE TITLE
y_iterate problem

### ISSUE THREAD

**Issue by PT-Player** [pawn-lang/YSI-Includes#74]
Hi,

Some times the [iterator Player](https://github.com/Misiur/YSI-Includes/blob/YSI.tl/YSI_Data/y_foreach/iterators.inc#L224) is giving false information when i use: Iter_Count(Player)

sometimes i have 10 players on server and he says i have 11.. I dont know if can be lag or something like that..

Any ideas?

**Comment by Y-Less** [comment]
NPCs?

**Comment by Crayder** [comment]
> NPCs?

But shouldn't Player be just players? Character is both players and NPC's.

**Comment by Y-Less** [comment]
I can't remember which way round it is anymore sorry.

**Comment by Y-Less** [comment]
Can you provide a simplified test case that shows this and can be re-run to narrow down the bug please?

**Comment by PT-Player** [comment]
```
/*
    - PT_Player
*/

#include <YSI\y_hooks>

new Text:PlayersOnline[2],
    playeronstr[50];

hook OnGameModeInit()
{
    // Textdraws...
}

hook OnPlayerConnect(playerid)
{
    format(playeronstr, sizeof(playeronstr), "Online: %i/%i", Iter_Count(Player), MAX_PLAYERS);
    TextDrawSetString(PlayersOnline[1], playeronstr);   

    foreach(new i : Player)
    {
        if(!(i == playerid))
        {
            TextDrawShowForPlayer(i, PlayersOnline[1]);
        }
    }
}

hook OnPlayerDisconnect(playerid, reason)
{
    format(playeronstr, sizeof(playeronstr), "Online: %i/%i", Iter_Count(Player), MAX_PLAYERS);
    TextDrawSetString(PlayersOnline[1], playeronstr);   
    TextDrawShowForAll(PlayersOnline[1]);

    TextDrawHideForPlayer(playerid, PlayersOnline[0]);
    TextDrawHideForPlayer(playerid, PlayersOnline[1]);
}

hook OnPlayerSpawn(playerid)
{
    if(!(xPlayer[playerid][Logado] == false || xPlayer[playerid][Estado] == P_STATE_CRIAR_PERSONAGEM))
    {
        TextDrawShowForPlayer(playerid, PlayersOnline[0]);
        TextDrawShowForPlayer(playerid, PlayersOnline[1]);
    }
}
```

take an look @Y-Less 

But this problem dont happends all the time only after some players log and logout...

I have no NPC's on server.

**Comment by PT-Player** [comment]
@Misiur @Y-Less @Crayder no idea?

**Comment by Crayder** [comment]
Is that include example before or after you include y_iterate?

**Comment by Y-Less** [comment]
Without a consistent reproducible test case, there's no way to narrow down the problem or verify that it is fixed.  We need a simple y_testing "Test:" function that demonstrates the problem every time I'm afraid.

**Comment by PT-Player** [comment]
@Crayder the include order is: y_iterate -> My systems..

@Y-Less I gone try do that..

**Comment by Crayder** [comment]
Is it only the disconnect and connect hooks that are affected? Also, what are your expected results and what was the output for you (specifically for you since we're blaming y_iterate, if I tested that code I could possibly have different results)? Are these the only times the count is wrong?

**Comment by Y-Less** [comment]
Is it only `Player` that is wrong?  It is a bit more complex than normal iterators.

**Comment by PT-Player** [comment]
@Crayder the code is above i send, sometimes on server are 5 players but on textdraws show 6 ( just an exemple ).

@Y-Less yes i only saw this problem in the `Player` at the moment.

**Comment by Y-Less** [comment]
Is the problem the `Iter_Count` call in `OnPlayerDisconnect`?

**Comment by PT-Player** [comment]
Yes looks like that You have 4 players on Server and sometimes one disconnect and keeps showing 4..

You have the code above

**Comment by IstuntmanI** [comment]
Confirmed. I saw this is happening only when two+ players are connecting/disconnecting almost at the same time. It may be because of the Iter_Remove timer.

**Comment by Y-Less** [comment]
Yes, that would explain it actually.  If they managed to connect between the last one disconnecting and being removed from the iterator.  If that is the case, it is easy to fix.

**Comment by Y-Less** [comment]
Actually, no, because:

https://github.com/Misiur/YSI-Includes/blob/f1b57c293a09a9f3c43e36f6b6cf7798efb5f45c/YSI_Data/y_foreach/iterators.inc#L549-L552

**Comment by IstuntmanI** [comment]
I didn't mean that. I think it is something like this:

``` pawn
OnPlayerConnect( 5 );
OnPlayerConnect( 10 );
OnPlayerConnect( 15 );
OnPlayerConnect( 20 );
OnPlayerConnect( 25 );

printf( "Players: %d", Iter_Count( Player ) ); // should show 5, and it does (if I remember correctly)

OnPlayerDisconnect( 25, 0 );
OnPlayerDisconnect( 10, 0 );
OnPlayerDisconnect( 15, 0 );

printf( "Players: %d", Iter_Count( Player ) ); // should show 2, but AFAIK, it is showing a wrong value, probably still 5, because the timer wasn't called yet
```

I think this happens only when the disconnecting is really fast. (as you know, that 0 delay from the Iter_Remove timer is actually at the next process tick, or a even bigger difference) I didn't test the code, but as far as I can remember, this is what happens. If I use Iter_Count( player ) after that code (in a timer or other callback) it is showing the right value: 3.

---

Take a look at this bug: http://forum.sa-mp.com/showpost.php?p=3722109&postcount=22 . In my server I have a textdraw with 5 lines at the bottom of the screen which shows the last few events (mainly disconnects and connects). We had like 34 players when that mass timeout occured and then we got to ~17, so the 5 lines were full of the last 5 disconnect messages, in which I also show the remaining players, on each line. Each of those 5 lines had something like _[Players: 23/50]_ instead of it decreasing with each line, when the real count after the last disconnect was actually 17 (it wasn't showing 34, because I guess it got to decrease the count between the timeouts). But then, when a player connected it was showing _[Players: 18/50]_, like it should, so the count isn't bugged.

It must be because of that timer. I don't see anything else which would affect it.

**Comment by Y-Less** [comment]
Ahh, I see.  So if only shows the wrong count at certain points, but doesn't actually cause long-term damage by like adjusting who it thinks is connected permanently.  Unfortunately, I'm not sure of any good way to fix that - the count will still be off because the iterator will still be off.  The problem is that doing `Kick` inside a `foreach` loop will cause `OnPlayerDisconnect` to be called immediately, and remove items from the iterator while inside the iterator.  The timer was added to move the removal to outside the loop, and without a better solution you will always have one problem or the other.

**Comment by Y-Less** [comment]
The only thing I can think of is to hook `Kick` and do the timer removal if that was used, and don't if it wasn't.  Since `ShowPlayerDialog` was commented out, that might be the only function hook in YSI...
